### ISSUE TITLE
Changes in the directory structure of NVE dir(s)

### ISSUE THREAD

**Issue by tunnckoCore** [ehmicky/nve#11]
**Which problem is this feature request solving?**

Doesn't work with `global-dirs` and probably other such libraries.

**Describe the solution you'd like**

Currently the `node` bin is the `nve/<VERSION>/node`, while it's supposed to be inside a `/bin` directory inside the version. E.g. `nve/12.13.0/bin/node`.

Currently, with NVM

```
{
  npm: {
    prefix: '/home/charlike/.nvm/versions/node/v12.13.0',
    packages: '/home/charlike/.nvm/versions/node/v12.13.0/lib/node_modules',
    binaries: '/home/charlike/.nvm/versions/node/v12.13.0/bin'
  }
}
```

Currently, with NVE

```
{
  npm: {
    prefix: '/home/charlike/.cache/nve',
    packages: '/home/charlike/.cache/nve/lib/node_modules',
    binaries: '/home/charlike/.cache/nve/bin'
  }
}
```

The problem comes because `global-dirs` is doing double `path.dirname` on `process.execPath` which for NVE is `/home/charlike/.cache/nve/12.13.0/node` and that's why we are getting `/home/charlike/.cache/nve`

**Describe alternatives you've considered**

Alternative would be to go to `global-dirs` and suggest addition of one `if` there. But I don't think that is the correct way and place to fix this. Instead, this module should adjust to the convention.

**Can you submit a pull request?**

Yes, probably.

**Comment by ehmicky** [comment]
Hi @tunnckoCore,

First of all, thanks a lot for pointing this out. This is indeed a use case that I completely forgot and I'm glad you're mentioning it. The answer to this is a little complex, so please bear with me :)

## nvm vs nve

`nve` is not meant to replace `nvm`. I use both on my machine. I use `nvm` to change the Node.js version for my entire machine or for a specific project. I use `nve` for one-off quick commands such as `nve 8 npm test` (checking projects works on Node 8), `nve 8 10 12 npm run benchmark` (benchmarking among several Node versions) or `nve 8 node -p "..."` (check if an older Node.js version has a specific syntax or feature).

I've added a section to the `README` [explaining the difference](https://github.com/ehmicky/nve#difference-with-nvm).

## Problem

The underlying problem is that `nve` does not currently work well with `npm`, `yarn` and `pnpm`. `global-dirs` is just replicating the logic from those tools, which is why it fails too. 

`nve` is meant to download and execute a one-off/temporary `node` binary. It is not meant to replace the global `node` which must be downloaded and managed separately (for example with `nvm`). As such global binaries should be installed alongside the global `node` (e.g. in the `nvm` directory) not alongside `nve` cache directory. For example `nve` does not intend to modify the `PATH` environment variable. 

Now the issue is that this model breaks some of the assumptions of `npm`, `yarn` and `pnpm`. Namely those tools assume that the node executable (`process.execPath`) is in the same parent directory as the global binaries. In the case of `nve`, those are separate.

## Solution 1

One possible solution would be to install `nve` in the same directory as the global `node`, so it inherits the correct behavior. For example, instead of `/home/ehmicky/.nve/cache/13.1.0/node`, installing in `/home/ehmicky/.nvm/versions/node/v12.0.0/bin/node13.1.0`. 

This solution does not work though because this would require the Node.js binaries to be downloaded again each time you switch to a different Node.js version with `nvm`.

## Solution 2.1

Another solution would be for `nve` to set the `PREFIX` environment variable to the global `node` directory. That environment variable is used by `npm`, `yarn` and `pnpm`.

[`global-dirs`](https://github.com/sindresorhus/global-dirs) can be used to find the correct value.

## Solution 2.2

At the moment, the `node` executable is saved in `/home/ehmicky/.nve/cache/VERSION/node` on Unix. It would be safer to save it to `/home/ehmicky/.nve/cache/VERSION/bin/node` instead so that other tools that rely on `process.execPath` with the typical directory structure work properly. 

On Windows, the directory structure is flat, so no issues there.

## PR

I am going to work on two PRs implementing `2.1` and `2.2` if the solution sounds good to you.

**Comment by ljharb** [comment]
nvm has `nvm exec and `nvm run` for those entire use cases; how does nve differ from those?

**Comment by tunnckoCore** [comment]
> nve is not meant to replace nvm

Sure. But it can easily be total replacment.

> Solution 1

Yea, definitely not the way.

> Solution 2.1

I don't it's necessary to set `PREFIX`. Just put the downloaded node bin in `/bin` dir.

> I am currently working on two PRs implementing

Overall it's okay, yea.

**Comment by ehmicky** [comment]
@ljharb, first of all, thanks a lot for all the work you put into `nvm`, this is an amazing project! Just to be clear, `nve` is meant to be used alongside `nvm`, not replace it.

Differences with `nvm exec` and `nvm run` include:
   - can run multiple Node.js versions at once
   - can be run programmatically
   - works on Windows. No need to run as Administrator.
   - does not require Bash
   - is installed as a Node module (as opposed to a Bash installation script downloaded with curl)
   - does not need a separate installation step for each Node version

**Comment by ehmicky** [comment]
@tunnckoCore 

> Sure. But it can easily be total replacment. Someone may want to use it that way.

Replacing the global Node.js binary is a different problem space with very different constraints. For example:
  - it requires modifying the `PATH` environment variable
  - because of this, since child processes cannot modify their parent environment, it requires hooking up into the current shell initialization script (e.g. `bashrc`). This means the code must be shell-specific.
  - and much more

Those problems are perfectly solved by `nvm` (and related tools) so there's no need for `nve` to try to reinvent the wheel. On the other hand, I personally think we are missing a tool to run specific Node.js version:
  - multiple ones, serially or in parallel
  - friendly to programmatic usage
  - installed a a Node module
  - with no separate installation steps.

This is the problem `nve` is aiming at solving.

> I don't it's necessary to set PREFIX. Just put the downloaded node bin in `/bin` dir.

Could you explicit what you mean here? Do you mean the global `/bin` directory or `~/.nve/cache/VERSION/bin/`? The global `/bin` directory requires root permissions

**Comment by ljharb** [comment]
@ehmicky thanks for the explanation! Altho “ can run multiple Node.js versions at once”, do you mean it can run my command in node 6 and 8, say, simultaneously, and show me both outputs?

**Comment by ehmicky** [comment]
@ljharb `nve` can run several Node.js versions either serially or in parallel (see [here](https://github.com/ehmicky/nve/blob/master/README.md#examples-multiple-versions) for examples). It can also abort on the first error or not. The outputs of all Node.js versions is shown in order (even when they were run in parallel). 

Running several Node.js versions in parallel then showing their output in order is not completely trivial for several reasons, including the fact that the child processes are in a non-interactive TTY even if the command was run in an interactive TTY. For example colors might be missing in the output. `nve` includes several pieces of logic for this feature (code [here](https://github.com/ehmicky/nve/blob/master/src/parallel.js)).

**Comment by ehmicky** [comment]
@tunnckoCore this issue should be solved by the latest release [`7.3.0`](https://github.com/ehmicky/nve/releases/tag/7.3.0), could you please confirm it now works?

**Comment by tunnckoCore** [comment]
> Do you mean the global /bin directory or ~/.nve/cache/VERSION/bin/

Yup the bin inside the version dir.

> by the latest release 7.3.0,

Sure. I'll try.

**Comment by tunnckoCore** [comment]
Mm. Not exactly. Things got even more messy. :D 

the `file.js`

```js
const dirs = require('global-dirs');

console.log(dirs);
```

With node 12.13

```
❯ node file.js
{
  "npm": {
    "prefix": "/home/charlike/.nvm/versions/node/v12.13.0",
    "packages": "/home/charlike/.nvm/versions/node/v12.13.0/lib/node_modules",
    "binaries": "/home/charlike/.nvm/versions/node/v12.13.0/bin"
  },
  "yarn": {
    "prefix": "/home/charlike/.config/yarn",
    "packages": "/home/charlike/.config/yarn/global/node_modules",
    "binaries": "/home/charlike/.config/yarn/global/node_modules/.bin"
  }
}
```

With `nve@7.3`

```
❯ nve 10 node file.js 
{
  "npm": {
    "prefix": "/home/charlike/.nvm/versions/node/v12.13.0",
    "packages": "/home/charlike/.nvm/versions/node/v12.13.0/lib/node_modules",
    "binaries": "/home/charlike/.nvm/versions/node/v12.13.0/bin"
  },
  "yarn": {
    "prefix": "/home/charlike/.nvm/versions/node/v12.13.0",
    "packages": "/home/charlike/.nvm/versions/node/v12.13.0/global/node_modules",
    "binaries": "/home/charlike/.nvm/versions/node/v12.13.0/global/node_modules/.bin"
  }
}
```

So, yes, the `npm` part is correct now, but `yarn` part is wrong. It's probably because the fact that you are touching the `PREFIX` as suggested?

**Comment by ehmicky** [comment]
Thanks for letting me know.

This should now be fixed with `nve` [`7.3.1`](https://github.com/ehmicky/nve/releases/tag/7.3.1).

**Comment by tunnckoCore** [comment]
Works! <3
